### PR TITLE
Skip HEREDOCs when parsing perl virtual Provides

### LINKED ISSUE #1106: perl.prov mishandles FatPacked modules
Some modules use [Module::FatPack](https://metacpan.org/pod/Module::FatPack) to embed dependencies in a pm file themselves. Functionally, this packs them into HEREDOCs, and expands the code at runtime. I've only seen this so far in [App::cpanminus](https://metacpan.org/pod/App::cpanminus), in this file:
https://metacpan.org/source/MIYAGAWA/App-cpanminus-1.7044/lib/App/cpanminus/fatscript.pm

The end result is that perl.prov generates Provides for each packed module, even though it's not actually provided outside the execution of the parent package.

### PR DESCRIPTION
**Pull request by bwarden**
As noted in particular with App::cpanminus, a lot of perl virtual
provides are created that the package *doesn't actually provide*. This
package uses Module::FatPack to embed required libraries that may not be
already installed on the system, which defines them in HEREDOCs for
later use.

The only other example where I found package code completely within a
HEREDOC was in Mail::Sender, and that code is intended to be used only
privately within Mail::Sender itself. This seems like a valid set of
code to explicitly ignore.

This is an attempt to make the parser identify, then skip, HEREDOCs when
looking for provided packages.

Fixes #1106

### CODE REVIEW

**Comment by Conan-Kudo** [comment]
@ppisar What do you think of this?

**Comment by ppisar** [comment]
Perl is a dynamic language and it's impossible to catch all wanted instances
without introducing false positives and negatives. The more cases you will
cover the more complicated the code will be. While the idea behind the patch
is good in general, I won't asses whether it introduces regressions or not.
Both functional and perfomance-wise.

**Comment by bwarden** [comment]
> Perl is a dynamic language and it's impossible to catch all wanted instances without introducing false positives and negatives. The more cases you will cover the more complicated the code will be.

I absolutely agree. The only thing that can dependably parse Perl is `perl`, not any Perl code. I won't pretend this patch is perfect; in fact I'm quite confident it won't even handle all valid HEREDOCs correctly, simply because there are so many variant cases and it would make this code far more complex to attempt to catch them.

What I do have, though, is at least some data. I ran this against our corpus of shipped Perl modules, and compared the output. The only differences were:
- Eliminated the pile of false matches from App::cpanminus
- Eliminated perl(Mail::Sender::DBIO) and perl(Mail::Sender::IO), which are only used internally by Mail::Sender itself, not exposed.
- Eliminated perl(Template::Perl) from Template::Toolkit, which also appears to be used only internally
- Eliminated perl(warnings) from mod_perl

**Comment by pmatilai** [comment]
@ppisar , that changes like this have potential for regressions is obvious even to a perl-illiterate like myself. Let me rephrase that question: would *you* merge this? If it's left up to me, I'll hit the merge button in a day or two if only because I don't know any better :smile:

**Comment by bwarden** [comment]
[This project](https://metacpan.org/release/rpm-build-perl) aims to provide replacements that hook into the compiler backend of Perl and could theoretically give better answers. I had to apply a [patch](https://rt.cpan.org/Ticket/Attachment/1731178/930892/rpm-build-perl-0.82-Port-to-OpSIBLING-like-macros-required-since-Perl-5..patch) from a [bug report](https://rt.cpan.org/Public/Bug/Display.html?id=117350). It outputs dependencies differently

>     Convert module name to conventional dependency name, e.g. File::Find -> perl(File/Find.pm). Note that this differs from RedHat conventional form perl(File::Find).

but that's a minor thing to change. Running it against App::cpanminus::fatscript:
```
$ perl.prov /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm
# saw $VERSION in package App::cpanminus at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 25 (SKIP)
# saw $VERSION in package App::cpanminus::script at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 423 (SKIP)
# saw $VERSION in package CPAN::DistnameInfo at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 3478 (SKIP)
# saw $VERSION in package CPAN::Meta at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 3688 (SKIP)
# saw $CPAN::Meta::Check::VERSION at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 4843 (SKIP)
# saw $VERSION in package CPAN::Meta::Converter at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 4978 (SKIP)
# saw $VERSION in package CPAN::Meta::Feature at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 6620 (SKIP)
# saw $VERSION in package CPAN::Meta::History at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 6773 (SKIP)
# saw $VERSION in package CPAN::Meta::Merge at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 7091 (SKIP)
# saw $VERSION in package CPAN::Meta::Prereqs at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 7391 (SKIP)
# saw $VERSION in package CPAN::Meta::Requirements at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 7816 (SKIP)
# saw $VERSION in package CPAN::Meta::Spec at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 8927 (SKIP)
# saw $VERSION in package CPAN::Meta::Validator at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 10160 (SKIP)
# saw $CPAN::Meta::YAML::VERSION at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 11375 (SKIP)
# saw $VERSION in package Exporter at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 12333 (SKIP)
# saw $VERSION in package Exporter::Heavy at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 12922 (SKIP)
# saw $VERSION in package File::pushd at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 13177 (SKIP)
# saw $VERSION in package HTTP::Tiny at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 13482 (SKIP)
# saw $JSON::PP::VERSION at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 15704 (SKIP)
# saw $JSON::PP::IncrParser::VERSION at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 17114 (SKIP)
# saw $VERSION in package Module::CPANfile at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 18526 (SKIP)
# saw $VERSION in package Module::Metadata at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 19207 (SKIP)
# saw $VERSION in package Parse::CPAN::Meta at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 20220 (SKIP)
# saw $VERSION in package Parse::PMFile at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 20576 (SKIP)
# saw $VERSION in package String::ShellQuote at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 21497 (SKIP)
# saw $VERSION in package local::lib at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 21773 (SKIP)
# saw $VERSION in package local::lib at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 21774 (SKIP)
# saw $VERSION in package parent at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 23207 (SKIP)
# saw $VERSION in package version at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 23355 (SKIP)
# saw $VERSION in package version::regex at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 23477 (SKIP)
# saw $VERSION in package version::vpp at /usr/lib/perl5/vendor_perl/5.30.1/App/cpanminus/fatscript.pm line 23716 (SKIP)
perl(App/cpanminus/fatscript.pm)
```

**Comment by bwarden** [comment]
Oh, @ppisar, I guess you wrote that patch :D

**Comment by pmatilai** [comment]
Okay, since nobody is screaming "no you can't do that", I'll just say thanks for the patch and hit the merge button.
